### ISSUE TITLE
Best practices with ppc64le/aarch64 migration

### ISSUE THREAD

**Issue by jakirkham** [conda-forge/conda-forge.github.io#870]
For a while now we have been using Azure with QEMU to virtualize architectures we didn’t have easy access to. In a few cases we have used Travis CI to do ppc64le builds on native infrastructure. Recently Drone came online and that allows us to do native aarch64.

It was my understanding that we should try to use Azure wherever possible for these architectures as they have generously provided us resources and only use native builds when there is an issue with emulation.

Based on a recent discussion with @isuruf, it seems that my knowledge here may be out-of-date. Given the number of feedstocks involved and the number of people that will be involved, I think it is very important that we establish best practices for how we want to handle builds for these architectures and communicate that clearly to people (IOW document). It’s totally possible that the best practices have been established by those close to the migration, but the communication thereof does not appear to be happening. Perhaps even these are documented and we are just lacking visibility.

Anyways am opening this issue to make sure we have firmed up what those best practices are, document them, and make sure that info is visible.

Apologies if I’m just missing lots of things here. Figured that if I was missing things, there are many maintainers who are also missing them, which seems worth correcting if nothing else.

cc @conda-forge/core

**Comment by isuruf** [comment]
For non-x86 architectures, it's best practice to do something like,

```yaml
provider:
   linux_ppc64le: default
```
instead of

```yaml
provider:
   linux_ppc64le: azure
```
or

```yaml
provider:
   linux_ppc64le: travis
```

This way, we can switch the provider in conda-smithy from emulated to native if support becomes available in conda-smithy.

As for resources, Travis-CI and Drone are doing fine at the moment and we can always switch if that becomes a problem if the provider is named `default`

**Comment by isuruf** [comment]
> It was my understanding that we should try to use Azure wherever possible for these architectures as they have generously provided us resources and only use native builds when there is an issue with emulation.

We should always use native builds as emulation can be okay for some versions and not for some versions. When there's a version update, emulation job might break just because of some bug in emulation, and the maintainers might ignore that platform.

**Comment by CJ-Wright** [comment]
I think this is the bot's current practice https://github.com/regro/cf-scripts/blob/master/conda_forge_tick/migrators.xsh#L1011

**Comment by hmaarrfk** [comment]
What is our time limit on drone?

**Comment by isuruf** [comment]
1 hour

**Comment by hmaarrfk** [comment]
And how many parallel builds do we get? The issue I ran into with shippable was that they didn't auto cancel jobs when an update was provided, and that started to create large queues that would grow indefinitely. 

I think it might be best to start a build with azure, and only move to native once problems occur.

**Comment by hmaarrfk** [comment]
Allowing people to add things like the make dependency, or missing cdts without hogging up the drone resources.

**Comment by isuruf** [comment]
We agreed on using native as much as possible at https://github.com/conda-forge/conda-smithy/pull/1023#discussion_r272854099. If you want to change it, please open an issue in conda-smithy and give concrete reasons.

**Comment by hmaarrfk** [comment]
Ok

**Comment by h-vetinari** [comment]
Not sure if this worth its own issue, but w.r.t. drone, it would be very nice to:
* be able to restart failed stages separately - it feels very wasteful to restart 4 long-running builds (e.g. scipy) in an already backlogged queue when only one build failed
* create a sub-queue for the larger of the two drone machines, where feedstocks that hit the timeout on the smaller machine can explicitly opt in (rather than having to restart if they hit a small machine for any of the jobs)

I remember from previous discussions (not sure where anymore) that - paraphrasing - "nothing about drone can be changed". If that's the case then so be it, but I just wanted to bring it up because I've been trying for almost 2 days to rebuild that one last missing aarch-package of scipy, and felt I should comment.

**Comment by hmaarrfk** [comment]
I think crosscompiling is becoming more and more attractive.

If instructions aren't clear on how to do that, we should improve them on the documentation website.

**Comment by h-vetinari** [comment]
The instructions are not the issue (to me at least). Cross-compilation works by and large, but where it doesn't, the set up on drone is more wasteful than strictly necessary for an already constrained queue (currently several hours of waiting time until a job starts), through the three points I mentioned.

For cross-compilation, builds still time out due to running through emulation, despite the ~6-fold time limit, see for example https://github.com/conda-forge/scipy-feedstock/pull/179. And more serious problems like certain errors only appearing in emulation (e.g. https://github.com/scipy/scipy/issues/14560), which makes cross compilation impossible for affected libraries.

**Comment by hmaarrfk** [comment]
Cross compiling is different than using qemu.

The last attempt I know of that tried to cross compile was
https://github.com/conda-forge/scipy-feedstock/pull/135

Issues regarding numpy distutils may have been solved recently

https://labs.quansight.org/blog/2021/07/moving-scipy-to-meson/

**Comment by h-vetinari** [comment]
> Cross compiling is different than using qemu.

Yeah, fair point, I was too loose in the usage of the term (I do understand the difference). What I don't like with cross-compilation is that the artefacts aren't actually tested in regular CI (unless we can link e.g. the azure-output with running the test suite on drone/travis...? that sounds unlikely though).

**Comment by hmaarrfk** [comment]
I guess i can understand the expanded use of the term.

**Comment by isuruf** [comment]
> What I don't like with cross-compilation is that the artefacts aren't actually tested in regular CI

They are tested if `test_on_native_only: true` is removed. See https://github.com/conda-forge/conda-smithy/issues/1455
